sympy/matrices/matrices.py

```diff
@@ -3986,3 +3986,117 @@ def classof(A, B):
 def a2idx(j, n=None):
     from sympy.matrices.common import a2idx as a2idx_
     return a2idx_(j, n)
+
+
+def randMatrixBigDiag(rows, cols=None, lo=0, hi=100, symmetric=False):
+    """Return a matrix filled with random values such that no
+    value is greater than the values on the diagonals in the same
+    row or column containing the value.
+
+    Examples
+    ========
+
+    >>> randMatrixBigDiag(4)
+    Matrix([
+    [20,  6,  7, 12],
+    [ 0, 36, 19, 23],
+    [17, 15, 43, 20],
+    [10,  5,  3, 45]])
+    >>> randMatrixBigDiag(4, symmetric=True)
+    Matrix([
+    [56, 54, 17, 30],
+    [54, 87,  5, 17],
+    [17,  5, 35, 16],
+    [30, 17, 16, 34]])
+    >>> randMatrixBigDiag(4, 5)
+    Matrix([
+    [55, 28, 31, 23],
+    [29, 32,  3,  6],
+    [12, 16, 75, 42],
+    [30,  1, 14, 56]])
+    >>> randMatrixBigDiag(4, hi=10, symmetric=True)
+    Matrix([
+    [6, 5, 1, 2],
+    [5, 7, 2, 2],
+    [1, 2, 9, 1],
+    [2, 2, 1, 5]])
+    """
+    from sympy.matrices.dense import diag
+    from random import randint
+    cols = rows if cols is None else cols
+    d = [randint(lo, hi) for i in range(min(rows, cols))]
+    A = diag(*d)
+    if symmetric:
+        if not len(set(A.shape)) == 1:
+            raise ValueError('only a square matrix can be symmetric')
+        for i in range(A.rows):
+            for j in range(i + 1, A.cols):
+                A[j,i] = A[i,j] = randint(0, min(d[i], d[j]) - 1)
+    else:
+        for i in range(A.rows):
+            for j in range(A.cols):
+                if i != j:
+                    A[i,j] = randint(0, min(d[i], d[j]) - 1)
+    return A
+
+def upper(a, diag=0):
+    """
+    Examples
+    ========
+
+    >>> upper(ones(4,2))
```

Comment:
There should probably be `from sympy.matrices.matrices import upper` and `from sympy.matrices import ones` before this line. You can test the examples by invoking python and then copying these statements before doing anything else.